doc/threshold/SONiC Threshold feature spec.md

```diff
@@ -0,0 +1,589 @@
+
+The supported buffers for threshold are:
+
+1. Ingress port priority group
+	1. Shared - Specifies the shared buffer usage on a priority group.
+	2. Headroom - Specifies the additional buffer limit that can be utilized when the shared limits are exhausted in the case where flow control is enabled. It accounts for the round trip time of data to be received once flow control is asserted. 
+2. Egress port queues
+	1. Unicast 
+	2. Multicast
+ 
+
+# 3 Design
+## 3.1 Overview
+
+![Threshold architecture](images/Threshold_arch.jpg)
+
+The above diagram illustrates the architecture of the threshold feature within SONiC. The call flow sequence is listed below:
+
+1. CLI takes threshold configuration and populates it into THRESHOLD\_TABLE of config DB.
+2. The ThresholdOrch component of Orchagent in SWSS docker picks up the config DB updates on THRESHOLD\_TABLE and converts the config into SAIREDIS calls. ASIC\_DB is updated by SAIREDIS and SYNCD gets notified of incoming updates.
+3. SYNCD invokes the SAI APIs for configuring the threshold. 
+4. SAI configures the ASIC with threshold configuration.
+5. ASIC notifies SAI of a threshold breach notification.
+6. SAI collates the breach information into a protobuf message and sends it out to the configured collector (in this case, ThresholdMgr) on a socket.
+7. ThresholdMgr parses the threshold breach protobuf message and populates the information into COUNTERS\_DB.
+8. The data available in COUNTERS\_DB is used by CLI for showing breach information.  
+
+Some salient points are listed below:
+
+1. ThresholdMgr currently only handles the breach event protobuf processing. It can be extended to support handling of dynamic port breakout configuration etc. in future.
+2. ThresholdOrch subscribes to notifications on THRESHOLD\_TABLE of config DB. Since the configuration is in-line with SAI configuration, the orch agent directly picks this config up from config DB.
+3. TAM docker currently only hosts ThresholdMgr. It is intended that new TAM applications would be hosted in this container going forward.
+
+## 3.1.1 ThresholdMgr
+
+ThresholdMgr running in the TAM container handles the protobuf parsing of the threshold breach report data. SAI is configured by ThresholdOrch to send the threshold breach event notification to a local collector i.e ThresholdMgr.
```

Comment:
since tam is part of SAI spec, then why threshold mgr to part of syncd?

Reply:
Hi,

Thresholdmgr is running on a separate container, TAM. The thresholdmgr receives the protobuf from SAI directly by-passing the syncd and orch layers. This is done since the protobuf data from SAI does not need any translation and can be processed directly by thresholdmgr.

Thresholdmgr will be expanded in future to handle multiple other cases  like dynamic port breakout which would not be part of syncd. Hence, thresholdmgr is setup as part of TAM container.